### ISSUE TITLE
Consider PreserveDependencyAttribute to help linker

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#30902]
We currently use ILLinkTrim.xml files to express APIs, whether public or non-public, that should not be trimmed away by the linker, whether during the first trimming phase when we build the assemblies or the second optional trimming phase when an app is deployed with trimming.

However, these declarations are causing us to keep around a lot more state than is necessary, for two reasons:
1. The presence of such an .xml file is forcing the binary to be preserved, even if nothing from it ends up being used.
2. It doesn't provide the ability to say "only keep XYZ if ABC is used".

For example, in System.Data.Common, we have an ILLinkTrim.xml file that says the DataTableTypeConverter's .ctor() should be preserved.  That's because the DataView.Table property is annotated with a TypeConverterAttribute:
https://github.com/dotnet/corefx/blob/19b304f7815894b13cb61e87e1c9eac49a474c7e/src/System.Data.Common/src/System/Data/DataView.cs#L473
and the type converter infrastructure needs access to that ctor to instantiate it via reflection.  However, because that .ctor is mentioned in the .xml file, it'll never be trimmed away, even if DataView.Table itself isn't used.  And DataTableTypeConverter's ctor references DataTable, which in turn references a whole bunch of stuff.

The same goes for other assemblies, e.g. System.Text.Json, System.Private.DataContractSerialization, etc.

We should consider instead exposing and using the PreserveDependencyAttribute that the linker already has some support for (though coreclr and corefx are currently using too old a version), e.g.
```C#
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Field, AllowMultiple = true, Inherited = false)]
    internal sealed class PreserveDependencyAttribute : Attribute
    {
        public PreserveDependencyAttribute(string? memberSignature)
        {
            MemberSignature = memberSignature;
        }

        public PreserveDependencyAttribute(string? memberSignature, string typeName)
        {
            MemberSignature = memberSignature;
            TypeName = typeName;
        }

        public PreserveDependencyAttribute(string? memberSignature, string typeName, string assemblyName)
        {
            MemberSignature = memberSignature;
            TypeName = typeName;
            AssemblyName = assemblyName;
        }

        public string? MemberSignature { get; set; }
        public string? TypeName { get; set; }
        public string? AssemblyName { get; set; }

        public string? Condition { get; set; }
    }
}
```

This, or something like it, has the ability to specify not just that something is needed, but why that thing is needed, in a way where if the reason for it needing to be there gets trimmed away, so too does the need.  It also has the benefit that the dependency is expressed in the code rather than in a separate asset.

cc: @sbomer, @jkotas, @marek-safar

**Comment by MichalStrehovsky** [comment]
In light of mono/linker#805 we should look for a different name:

Some proposed names from the issue:

* `HasDependencyOnAttribute`
* `ImplicitDependencyAttribute`
* `MemberDependencyAttribute`
* Throwing one more name `DynamicDependencyAttribute`

**Comment by vitek-karas** [comment]
I like the `DynamicDependency`.

**Comment by marek-safar** [comment]
This is more like `ExplicitDependencyAttribute` than `ImplicitDependencyAttribute` especially when used for dependencies channelled via native code

**Comment by MichalStrehovsky** [comment]
> especially when used for dependencies channelled via native code

Is this a customer scenario? It feels like the use cases where we need to preserve something because native code is essentially going to "reflect" on this are more our internal/Unity use cases and not something mainstream customers would hit. (I want to make sure we optimize the name for customer scenarios and not our internal use cases.)

It feels that the callbacks from native code still fit the "dynamic" name because they're all essentially just invoked through a different kind of reflection (get looked up using string names, called in a funny way, etc.).

But I don't have a very strong preference for this attribute in particular.

**Comment by terrajobst** [comment]
* It seems wild cards are also supported, which raises a few questions:
	- Does it include private and public APIs or just public?
	- How hard is it for upstream dependencies to implement (e.g. ILSpy or VS Find All Usages)
	- Do we need them? We'd like because it seems [there are places](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs#L278-L319) where we have a lot.
* The linker should at least warn when the strings can't be matched
* `typeName` and `assemblyName` are optional because they default to same and same assembly
* Looks like `memberSignature` should be non-nullable
* Condition is some specific string that is interpreted by the linker, today that's only `DEBUG` (which unfortunately doesn't map to the C# notion of compiling a debug build)

```C#
namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.Method |
                    AttributeTargets.Constructor |
                    AttributeTargets.Field,
                    AllowMultiple = true,
                    Inherited = false)]
    public sealed class DynamicDependencyAttribute : Attribute
    {
        public DynamicDependencyAttribute(string memberSignature)
        {
            MemberSignature = memberSignature;
        }

        public DynamicDependencyAttribute(string memberSignature, string typeName)
        {
            MemberSignature = memberSignature;
            TypeName = typeName;
        }

        public DynamicDependencyAttribute(string memberSignature, string typeName, string assemblyName)
        {
            MemberSignature = memberSignature;
            TypeName = typeName;
            AssemblyName = assemblyName;
        }

        public string MemberSignature { get; }
        public string? TypeName { get; }
        public string? AssemblyName { get; }
        public string? Condition { get; set; }
    }
}
```

**Comment by eerhardt** [comment]
Closed in #35387.
